### PR TITLE
`schemadiff`: multi-error in schema normalization

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

In this PR a `Schema`'s `normalize()` function, which also runs validations for the entire schema, tables and views, returns a potentially aggregated error rather than a single error.

The idea is to return as much information as possible about problems in the schema, we the user can have multiple action points they can handle at once. This is instead of receiving a single error, fixing it, trying again, receiving the next single error, etc.

Not all errors make sense to aggregate; we specifically aggregate errors that relate to view dependencies.

To top it all, `NewSchemaFromEntities()` returns whatever `Schema` object it may have processed, even if invalid, as opposed to returning `nil` on error.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/10203

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/schemadiff/schema.go [hunk 1]
```diff
@@ -330,7 +331,7 @@ func (s *Schema) normalize() error {
 	// Validate table definitions
 	for _, t := range s.tables {
 		if err := t.validate(); err != nil {
-			return err
+			return errors.Join(errs, err)
```
Comment (dbussink) [inline go/vt/schemadiff/schema.go:334]: In the above case of `s.ValidateViewReferences()` we don't want to return all gathered errors up to that point?

Comment (shlomi-noach) [reply go/vt/schemadiff/schema.go:334]: You're right, it's an oversight. Now both collecting the error along with all previous errors, _and_ also continuing the normalization process.

**Review by dbussink — APPROVED** [review]

File: go/vt/schemadiff/errors.go [hunk 2]
```diff
@@ -7,6 +7,39 @@ import (
 	"vitess.io/vitess/go/sqlescape"
 )
 
+// Wrapped is used to unwrap an error created by errors.Join() in Go 1.20
+type Wrapped interface {
+	Unwrap() []error
+}
+
+// Unwrap unwraps an error created by errors.Join() in Go 1.20, into its components
+func Unwrap(err error) []error {
```
Comment (rsajwani) [inline go/vt/schemadiff/errors.go:16]: This method will not be called right. u.Unwrap on line 27 is called on object level.

File: go/vt/schemadiff/schema_test.go [hunk 3]
```diff
@@ -153,6 +153,7 @@ func TestNewSchemaFromQueriesLoop(t *testing.T) {
 	)
 	_, err := NewSchemaFromQueries(queries)
 	assert.Error(t, err)
+	err = UnwrapOne(err)
 	assert.EqualError(t, err, (&ViewDependencyUnresolvedError{View: "v7"}).Error())
```
Comment (rsajwani) [inline go/vt/schemadiff/schema_test.go:157]: can we assert err length here as well .. to assure the numbers of errors are as expected

File: go/vt/schemadiff/errors.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/errors.go:16]: Right now no one is using this function... I added it for completeness.

File: go/vt/schemadiff/schema_test.go (hunk 3 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/schema_test.go:157]: `UnwrapOne()` s guaranteed to return a slice of at least one element.

Comment (shlomi-noach) [reply go/vt/schemadiff/schema_test.go:157]: And yes! I'll add that assertion anyway.

Comment (shlomi-noach) [reply go/vt/schemadiff/schema_test.go:157]: Rather, I'll create a `errors_test.go` with a bunch of unit tests.

Comment (shlomi-noach) [reply go/vt/schemadiff/schema_test.go:157]: Unit tests added

File: go/vt/schemadiff/errors.go [hunk 4]
```diff
@@ -7,6 +7,45 @@ import (
 	"vitess.io/vitess/go/sqlescape"
 )
 
+// Wrapped is used to unwrap an error created by errors.Join() in Go 1.20
+type Wrapped interface {
```
Comment (ajm188) [inline go/vt/schemadiff/errors.go:11]: this is more a structural comment; i'd suggest either renaming or relocating these types. right now you import `schemadiff.Wrapped` which, somehow, is for errors? i would expect from that name alone it's some sort of, well, wrapped diff structure (ditto the `Unwrap*` functions).

so, i'd either advocate for `schemadiff.WrappedError` or `vterrors.Wrapped` (or a new package!)

Comment (shlomi-noach) [reply go/vt/schemadiff/errors.go:11]: I agree. The location is suboptimal. I initially wanted to do this in `vterrors`, but `vterrors` has its own `Unwrap`. I'll create a new `errors` package.

Comment (shlomi-noach) [reply go/vt/schemadiff/errors.go]: Done, refactored!

**Review by ajm188 — APPROVED** [review]

**Pull request merged**